pkg/deploy/controller/deployerpod/controller_test.go

```diff
@@ -160,39 +166,128 @@ func TestHandle_podTerminatedFail(t *testing.T) {
 	}
 }
 
-// TestHandle_podTerminatedFailNoContainerStatus ensures that a failed
-// deployer pod with no container status results in a transition of the
-// deployment's status to failed.
-func TestHandle_podTerminatedFailNoContainerStatus(t *testing.T) {
-	var updatedDeployment *kapi.ReplicationController
-
+// TestHandle_cleanupDeploymentFailure ensures that clean up happens
+// for the deployment if the deployer pod fails.
+//  - failed deployment is scaled down
+//  - the last completed deployment is scaled back up
+func TestHandle_cleanupDeploymentFailure(t *testing.T) {
+	var existingDeployments *kapi.ReplicationControllerList
+	var failedDeployment *kapi.ReplicationController
+	// map of deployment-version to updated replicas
+	var updatedDeployments map[int]*kapi.ReplicationController
+
+	scaler := scalertest.FakeScaler{}
 	controller := &DeployerPodController{
 		deploymentClient: &deploymentClientImpl{
 			getDeploymentFunc: func(namespace, name string) (*kapi.ReplicationController, error) {
-				config := deploytest.OkDeploymentConfig(1)
-				deployment, _ := deployutil.MakeDeployment(config, kapi.Codec)
-				deployment.Annotations[deployapi.DeploymentStatusAnnotation] = string(deployapi.DeploymentStatusRunning)
-				return deployment, nil
+				return failedDeployment, nil
 			},
 			updateDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
-				updatedDeployment = deployment
+				if _, found := updatedDeployments[deployutil.DeploymentVersionFor(deployment)]; found {
+					t.Fatalf("unexpected multiple updates for the same deployment %d", deployutil.DeploymentVersionFor(deployment))
+				}
+				updatedDeployments[deployutil.DeploymentVersionFor(deployment)] = deployment
 				return deployment, nil
 			},
+			listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
+				return existingDeployments, nil
+			},
 		},
+		scaler: &scaler,
 	}
 
-	err := controller.Handle(terminatedPod())
+	type existing struct {
+		version         int
+		status          deployapi.DeploymentStatus
+		initialReplicas int
+		updatedReplicas int
+	}
 
-	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
+	type scenario struct {
+		// this is the deployment that is passed to Handle
+		version int
+		// this is the target replicas for the deployment that failed
+		desiredReplicas int
+		// existing deployments also include the one being handled currently
+		existing []existing
 	}
 
-	if updatedDeployment == nil {
-		t.Fatalf("expected deployment update")
+	scenarios := []scenario{
```

Comment:
Not a big deal but I think these would be easier to read if each `[]existing` entry were on its own line